### PR TITLE
[MRG+1] DeadlineCallback, early stopping for a fixed budget

### PR DESCRIPTION
**Pull request by yngtodd**
Creates a new `DeadlineCallback` class, an early stopping callback for fixed time budgets. When running optimizations on our clusters, we have fixed wall times for the jobs to be completed within. After pushing the clock a couple of times, I decided to write this class. 

After specifying the total amount of time you would like the optimization to run, the callback keeps track of the time taken for each iteration (exactly as in `TimerCallback`). It then decrements the iteration time from the total time, saving this as `time_remaining`. The early stopping occurs when the `time_remaining` is less than or equal to longest single iteration time. 

> Note: This code has some repetition from the `TimerCallback`. I originally tried to use multiple inheritance with both the `TimerCallback` and the `EarlyStopper` class, but ran into issues working around both parent classes' `__call__` methods. I also tried using composition as well a combination of inheritance and composition without much luck. If you guys have any thoughts on using multiple inheritance here, I am all ears!

### CODE REVIEW

**Comment by glouppe** [comment]
Looks like something useful! Could you add tests?

**Comment by betatim** [comment]
Ready for review? (change title to [MRG] if yes)

File: skopt/callbacks.py [hunk 1]
```diff
@@ -220,3 +220,37 @@ def _criterion(self, result):
...
+        self.total_time = total_time
+
+    def __call__(self, result):
+        elapsed_time = time() - self._time
+        self.iter_time.append(elapsed_time)
+        self._time = time()
+        return self._criterion(result)
+
+    def _criterion(self, result):
+        if len(result.x_iters) >= 1:
```
Comment (betatim) [inline skopt/callbacks.py]: should be `if result.x_iters` to test if the list contains something

File: skopt/callbacks.py [hunk 2]
```diff
@@ -220,3 +220,37 @@ def _criterion(self, result):
...
+            time_remaining = self.total_time - np.sum(self.iter_time)
+            return time_remaining <= np.max(self.iter_time)
+        else:
+            None
```
Comment (betatim) [inline skopt/callbacks.py]: missing `return`?

File: skopt/callbacks.py [hunk 3]
```diff
@@ -220,3 +220,37 @@ def _criterion(self, result):
...
+        finish within.
+    """
+    def __init__(self, total_time):
+        super(DeadlineStopper, self).__init__()
+        self._time = time()
+        self.iter_time = []
+        self.total_time = total_time
+
+    def __call__(self, result):
+        elapsed_time = time() - self._time
```
Comment (betatim) [inline skopt/callbacks.py]: You could move all this logic to `_cirterion` if you wanted to and then not have a `__call__`.

**Comment by betatim** [comment]
Thanks for putting up with my nitpicks :)

I restarted travis as it failed because of a timeout. Let's see if we can get a greenlight!

**Comment by glouppe** [comment]
Looks good to me. +1

**Comment by betatim** [comment]
🎆

**Comment by yngtodd** [comment]
Great, thanks everyone!

**Pull request merged**